packages/integrations/src/adguard-home/adguard-home-integration.ts

```diff
@@ -0,0 +1,156 @@
+
+    if (!stats.success) {
+      throw new Error(
+        `Failed to parse summary for ${this.integration.name} (${this.integration.id}), most likely your username or password is wrong: ${stats.error.message}`,
+      );
+    }
+
+    if (!status.success) {
+      throw new Error(
+        `Failed to parse summary for ${this.integration.name} (${this.integration.id}), most likely your username or password is wrong: ${status.error.message}`,
+      );
+    }
+
+    if (!filteringStatus.success) {
+      throw new Error(
+        `Failed to parse summary for ${this.integration.name} (${this.integration.id}), most likely your username or password is wrong: ${filteringStatus.error.message}`,
+      );
+    }
+
+    const blockedQueriesToday =
+      stats.data.time_units === "days"
+        ? (stats.data.blocked_filtering[stats.data.blocked_filtering.length - 1] ?? 0)
+        : stats.data.blocked_filtering.reduce((prev, sum) => prev + sum, 0);
+    const queriesToday =
+      stats.data.time_units === "days"
+        ? (stats.data.dns_queries[stats.data.dns_queries.length - 1] ?? 0)
+        : stats.data.dns_queries.reduce((prev, sum) => prev + sum, 0);
+    const countFilteredDomains = filteringStatus.data.filters
+      .filter((filter) => filter.enabled)
+      .reduce((sum, filter) => filter.rules_count + sum, 0);
+
+    return {
+      status: status.data.protection_enabled ? ("enabled" as const) : ("disabled" as const),
+      adsBlockedToday: blockedQueriesToday,
+      adsBlockedTodayPercentage: (queriesToday / blockedQueriesToday) * 100,
+      domainsBeingBlocked: countFilteredDomains,
+      dnsQueriesToday: queriesToday,
+    };
+  }
+
+  public async testConnectionAsync(): Promise<void> {
+    const username = super.getSecretValue("username");
+    const password = super.getSecretValue("password");
+
+    await super.handleTestConnectionResponseAsync({
+      queryFunctionAsync: async () => {
+        return await fetch(`${this.integration.url}/control/status`, {
+          headers: {
+            Authorization: `Basic ${Buffer.from(`${username}:${password}`).toString("base64")}`,
+          },
+        });
+      },
+      handleResponseAsync: async (response) => {
+        try {
+          const result = (await response.json()) as unknown;
+          if (typeof result === "object" && result !== null) return;
+        } catch (error) {
+          throw new IntegrationTestConnectionError("invalidJson");
+        }
+
+        throw new IntegrationTestConnectionError("invalidCredentials");
+      },
+    });
+  }
+
+  public async enableAsync(): Promise<void> {
+    const username = super.getSecretValue("username");
+    const password = super.getSecretValue("password");
+    const response = await fetch(`${this.integration.url}/control/protection`, {
+      method: "POST",
+      headers: {
+        "Content-Type": "application/json",
+        Authorization: `Basic ${Buffer.from(`${username}:${password}`).toString("base64")}`,
```

Comment:
Maybe you could add a private method that generates the authorization header so the logic of getting username, password and converting them to base64 has not to be done 4 times